sdk/program/src/instruction.rs

```diff
@@ -644,3 +645,112 @@ impl CompiledInstruction {
         &program_ids[self.program_id_index as usize]
     }
 }
+
+/// Use to query and convey the lengths of the inner instruction components
+/// when calling the syscall
+#[repr(C)]
+#[derive(Default, Debug, Clone, Copy)]
+pub struct InnerLengths {
+    pub data_len: u64,
+    pub accounts_len: u64,
+}
+
+/// Returns a inner instruction from the processed inner instruction list.
+///
+/// The processed inner instruction list is a reverse-ordered list of
+/// successfully processed inner instructions. For example, given the call flow:
+///
+/// A
+/// B -> C -> D
+/// B -> E
+/// B -> F
+///
+/// Then B's processed inner instruction list is: [F, E, D, C]
```

Comment:
I think it would be quite helpful to return the depth of the processed inner instructions as well, perhaps by changing the return type to `<Option<(usize, Instruction)>>`.

For example, `E` may want to confirm that `C` immediately preceded it and not care about what instructions `C` internally invoked (which may be `D` or otherwise depending on the arguments into `C`).  `E` currently has no way to figure that `D` was not at its invoke depth.

Reply:
+1

Reply:
Interesting, programs don't currently have any info on what depth they are being invoked.  We could either return a relative depth, or provide another syscall that provides the current depth.  The latter might be useful in other situations.

Reply:
> programs don't currently have any info on what depth they are being invoked

Native programs do via `InvokeContext::invoke_depth()`, it's been useful from time to time so exposing that into BPF programs via syscall as well seems quite nice.

Reply:
https://github.com/solana-labs/solana/pull/22515/commits/02a7f3d145f345dcb6def51d9a1a9075e159214e